File: app/src/main/java/com/example/a82105/login_ex/Retrofit/INodeJS.java
package com.example.a82105.login_ex.Retrofit;

import io.reactivex.Observable;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.POST;
import java.util.List;
import retrofit2.http.Body;

import android.os.Parcel;
import android.os.Parcelable;


public interface INodeJS {
    @POST("api/auth/register")
    @FormUrlEncoded
    Observable<String> registerUser(@Field("email") String email, @Field("password") String password, @Field("name") String name);

    @POST("api/auth/login")
    @FormUrlEncoded
    Observable<String> loginUser(@Field("email") String email, @Field("password") String password);



    // 장르에 따른 웹툰 가져오는 메서드
    @POST("api/recommend/get_webtoons")
    Observable<List<Webtoon>> getWebtoonsByGenre(@Body GenreRequest genreRequest);



    // 요청 클래스
    class GenreRequest {

        private String genre;

        public GenreRequest(String genre) {
            this.genre = genre;
        }

        public String getGenre() {
            return genre;
        }

        public void setGenre(String genre) {
            this.genre = genre;
        }
    }

    public class Webtoon implements Parcelable {
        String title;
        String thumbnail_link;

        public Webtoon(String title, String thumbnail_link) {
            this.title = title;
            this.thumbnail_link = thumbnail_link;
        }

        protected Webtoon(Parcel in) {
            title = in.readString();
            thumbnail_link = in.readString();
        }

        public static final Creator<Webtoon> CREATOR = new Creator<Webtoon>() {
            @Override
            public Webtoon createFromParcel(Parcel in) {
                return new Webtoon(in);
            }

            @Override
            public Webtoon[] newArray(int size) {
                return new Webtoon[size];
            }
        };

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(title);
            dest.writeString(thumbnail_link);
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getThumbnail_link() {
            return thumbnail_link;
        }

        public void setThumbnail_link(String thumbnail_link) {
            this.thumbnail_link = thumbnail_link;
        }
    }
}

File: app/src/main/java/com/example/a82105/login_ex/Retrofit/INodeJS.kt
package com.example.a82105.login_ex.Retrofit

import android.os.Parcel
import android.os.Parcelable
import io.reactivex.Observable
import retrofit2.http.Body
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.POST

interface INodeJS {
    @POST("api/auth/register")
    @FormUrlEncoded
    fun registerUser(
        @Field("email") email: String,
        @Field("password") password: String,
        @Field("name") name: String
    ): Observable<String>

    @POST("api/auth/login")
    @FormUrlEncoded
    fun loginUser(
        @Field("email") email: String,
        @Field("password") password: String
    ): Observable<String>

    // 장르에 따른 웹툰 가져오는 메서드
    @POST("api/recommend/get_webtoons")
    fun getWebtoonsByGenre(@Body genreRequest: GenreRequest): Observable<List<Webtoon>>

    // 요청 클래스
    data class GenreRequest(
        var genre: String
    )

    // Webtoon 클래스
    data class Webtoon(
        var title: String?,
        var thumbnail_link: String?
    ) : Parcelable {
        constructor(parcel: Parcel) : this(
            parcel.readString(),
            parcel.readString()
        )

        override fun writeToParcel(parcel: Parcel, flags: Int) {
            parcel.writeString(title)
            parcel.writeString(thumbnail_link)
        }

        override fun describeContents(): Int = 0

        companion object CREATOR : Parcelable.Creator<Webtoon> {
            override fun createFromParcel(parcel: Parcel): Webtoon {
                return Webtoon(parcel)
            }

            override fun newArray(size: Int): Array<Webtoon?> {
                return arrayOfNulls(size)
            }
        }
    }
}